File: QytetetJava/src/modeloqytetet/Qytetet.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package modeloqytetet;

import java.util.*;
import static java.util.Collections.shuffle;

/**
 *
 * @author sofiafernandezmoreno
 */
public class Qytetet {

    private ArrayList<Sorpresa> mazo = new ArrayList<>();
    private Tablero tablero = new Tablero();
    private EstadoJuego estado;

    private Dado dado = Dado.getInstance();
    private Sorpresa cartaActual;
    private ArrayList<Jugador> jugadores = new ArrayList();
    private Jugador jugadorActual;

    public final static int MAX_JUGADORES = 4;
    static final int MAX_SORPRESAS = 10;
    public static final int MAX_CASILLAS = 20;
    static final int PRECIO_LIBERTAD = 200;
    static final int SALDO_SALIDA = 1000;

    private static final Qytetet instance = new Qytetet();

    private Qytetet() {
        setCartaActual(null);
    } // Constructor

    public static Qytetet getInstance() {
        return instance;
    }

    ArrayList<Sorpresa> getMazo() {
        return mazo;
    }

    public Tablero getTablero() {
        return tablero;
    }

    Dado getDado() {
        return dado;
    }

    public Sorpresa getCartaActual() {
        return cartaActual;
    }

    private void setCartaActual(Sorpresa cartaActual) {
        this.cartaActual = cartaActual;
    }

    public ArrayList<Jugador> getJugadores() {
        return jugadores;
    }

    public void setJugadores(ArrayList<Jugador> jugadores) {
        this.jugadores = jugadores;
    }

    public Jugador getJugadorActual() {
        return jugadorActual;
    }

    public void setJugadorActual(Jugador jugadorActual) {
        this.jugadorActual = jugadorActual;
    }

    private void inicializarCartasSorpresa() {

        int randomNum1 = 3000 + (int) (Math.random() * 5000);
        int randomNum2 = 3000 + (int) (Math.random() * 5000);
        mazo.add(new Sorpresa("UHHH me convierto en Especulador", TipoSorpresa.CONVERTIRME, randomNum1));
        mazo.add(new Sorpresa("UHHH me convierto en Especulador", TipoSorpresa.CONVERTIRME, randomNum2));

        mazo.add(new Sorpresa("Un fan anónimo ha pagado tu fianza. Sales de la cárcel", TipoSorpresa.SALIRCARCEL, 0));
        mazo.add(new Sorpresa("Los killos te mandan de paseico", TipoSorpresa.IRACASILLA, 1));
        mazo.add(new Sorpresa("Lo siento, pero te quitamos un pizquito de tu saldo", TipoSorpresa.PAGARCOBRAR, 2));
        mazo.add(new Sorpresa("Nos debes dinerico señor, ¡a pagar!", TipoSorpresa.PORCASAHOTEL, 3));
        mazo.add(new Sorpresa("Killo que te toca pagar, $$$ , MALO MARIA", TipoSorpresa.PORJUGADOR, 4));
        mazo.add(new Sorpresa("Killo que ahora si te llevas el pellizquito", TipoSorpresa.PORJUGADOR, 5));
        mazo.add(new Sorpresa("Los killos te mandan de paseico un ratejo por si no te ha tocado", TipoSorpresa.IRACASILLA, 6));
        mazo.add(new Sorpresa("Fiuuuuuuu, que alegría que recibes dinerico", TipoSorpresa.PAGARCOBRAR, 7));
        mazo.add(new Sorpresa("Alegria señora que nos llevamos dinerico", TipoSorpresa.PORCASAHOTEL, 8));
        mazo.add(new Sorpresa("Te hemos pillado con chanclas y calcetines, lo sentimos,¡debes ir a la carcel!", TipoSorpresa.IRACASILLA, 9));

        shuffle(mazo);
    }

    private void inicializarTablero() {
        tablero = new Tablero();
    }

    private void inicializarJugadores(ArrayList<String> nombres) {
        jugadores = new ArrayList();
        for (String nombre : nombres) {
            jugadores.add(new Jugador(nombre) {
            });
        }
    }

    public void inicializarJuego(ArrayList<String> nuevosjugadores) throws Exception {
        inicializarJugadores(nuevosjugadores);
        inicializarCartasSorpresa();
        inicializarTablero();
        salidaJugadores();

    }

    public void siguienteJugador() {
        //Preguntar duda ¿utilizar metodo intentar salir carcel?

        int pos = jugadores.indexOf(jugadorActual);
        jugadorActual = jugadores.get((pos + 1) % jugadores.size());
        if (jugadorActual.getEncarcelado() == true) {
            estado = EstadoJuego.JA_ENCARCELADOCONOPCIONDELIBERTAD;
        } else {
            estado = EstadoJuego.JA_PREPARADO;
        }
    }

    void salidaJugadores() {
        for (Jugador j : jugadores) {

            //if (j.getCasillaActual().getTipo() != TipoCasilla.SALIDA) {
            j.setCasillaActual(this.tablero.obtenerCasillaNumero(0));//en el tablero la casilla 0 es la salida
            //}
        }
        jugadorActual = jugadores.get((int) (Math.random() * jugadores.size()));
        estado = EstadoJuego.JA_PREPARADO;
    }

    public ArrayList<Integer> obtenerPropiedadesJugador() {

        ArrayList<Integer> propiedadesJugadorNuevas = new ArrayList();

        for (TituloPropiedad p : jugadorActual.getPropiedades()) {
            for (Casilla s : tablero.getCasillas()) {
                if (s.getTitulo() == p) {
                    propiedadesJugadorNuevas.add(s.getNumeroCasilla());
                }
            }

        }
        return propiedadesJugadorNuevas;

    }

    public ArrayList<Integer> obtenerPropiedadesJugadorSegunEstadoHipoteca(boolean estadoHipoteca) {

        ArrayList<Integer> propiedadesJugadorNuevas = new ArrayList();
        for (TituloPropiedad p : jugadorActual.getPropiedades()) {
            for (Casilla s : tablero.getCasillas()) {
                if (s.getTitulo() == p && p.getHipotecada() == estadoHipoteca) {
                    propiedadesJugadorNuevas.add(s.getNumeroCasilla());
                }
            }
        }
        return propiedadesJugadorNuevas;
    }

    boolean jugadorActualEnCalleLibre() {
        boolean encontrado = false;
        for (TituloPropiedad p : jugadorActual.getPropiedades()) {
            if (jugadorActual.getCasillaActual().soyEdificable() && p.getPropietario() == null) {
                encontrado = true;
            }
        }
        return encontrado;
    }

    public boolean jugadorActualEncarcelado() {
        return jugadorActual.getEncarcelado();
    }

    public void jugar() {

        int valorDado = this.dado.Tirar();
        Casilla casillaPosicion = this.jugadorActual.getCasillaActual();
        Casilla casillaFinal = this.tablero.obtenerCasillaFinal(casillaPosicion, valorDado);
        mover(casillaFinal.getNumeroCasilla());

    }

    public void obtenerRanking() {
        Collections.sort(jugadores);
    }

    int obtenerSaldoJugadorActual() {
        return this.jugadorActual.getSaldo();
    }

    int tirarDado() {
        return this.dado.Tirar();
    }

    public int getValorDado() {
        return this.dado.getValor();
    }

    @Override
    public String toString() {
        return "Qytetet{" + "mazo=" + mazo + ", tablero=" + tablero + ", dado=" + dado + ", cartaActual=" + cartaActual + ", jugadores=" + jugadores + ", jugadorActual=" + jugadorActual + '}';
    }

    void actuarSiEnCasillaEdificable() {
        Casilla casilla;
        boolean deboPagar = jugadorActual.deboPagarAlquiler();
        if (deboPagar) {
            jugadorActual.pagarAlquiler();
            if (jugadorActual.getSaldo() <= 0) {
                setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);

            }
        }
        casilla = obtenerCasillaJugadorActual();
        if (estado == EstadoJuego.ALGUNJUGADORENBANCARROTA) {
            boolean tengoPropietario = ((Calle) casilla).tengoPropietario();
            if (tengoPropietario) {
                setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
            } else {
                setEstadoJuego(EstadoJuego.JA_PUEDECOMPRAROGESTIONAR);
            }
        }
    }

    public EstadoJuego getEstadoJuego() {
        return estado;
    }

    public Casilla obtenerCasillaJugadorActual() {
        return jugadorActual.getCasillaActual();
    }

    public void setEstadoJuego(EstadoJuego estado) {
        this.estado = estado;
    }

    void actuarSiEnCasillaNoEdificable() {
        setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        Casilla casillaActual = jugadorActual.getCasillaActual();
        if (casillaActual.getTipo() == TipoCasilla.IMPUESTO) {
            jugadorActual.pagarImpuesto();
        } else {
            if (casillaActual.getTipo() == TipoCasilla.JUEZ) {
                encarcelarJugador();
            } else if (casillaActual.getTipo() == TipoCasilla.SORPRESA) {
                cartaActual = mazo.remove(0);
                setEstadoJuego(EstadoJuego.JA_CONSORPRESA);
            }
        }
    }

    public void aplicarSorpresa() {
        setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);

        if (cartaActual.getTipo() == TipoSorpresa.SALIRCARCEL) {
            jugadorActual.setCartaLibertad(cartaActual);
        } else {
            mazo.add(cartaActual);
        }

        if (cartaActual.getTipo() == TipoSorpresa.PAGARCOBRAR) {
            jugadorActual.modificarSaldo(cartaActual.getValor());
            if (jugadorActual.getSaldo() < 0) {
                setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
            }
        } else if (cartaActual.getTipo() == TipoSorpresa.IRACASILLA) {
            int valor = cartaActual.getValor();
            boolean casillaCarcel = tablero.esCasillaCarcel(valor);
            if (casillaCarcel) {
                encarcelarJugador();
            } else {
                mover(valor);
            }

        } else if (cartaActual.getTipo() == TipoSorpresa.PORCASAHOTEL) {
            int cantidad = cartaActual.getValor();
            int numeroTotal = jugadorActual.cuantasCasasHotelesTengo();
            jugadorActual.modificarSaldo(numeroTotal * cantidad);
            if (jugadorActual.getSaldo() < 0) {
                setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
            }
        } else if (cartaActual.getTipo() == TipoSorpresa.PORJUGADOR) {
            for (Jugador j : jugadores) {
                if (j != jugadorActual) {
                    j.modificarSaldo(cartaActual.getValor());
                }
                if (j.getSaldo() < 0) {
                    setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                } else {
                    jugadorActual.modificarSaldo(-cartaActual.getValor());
                }
                if (jugadorActual.getSaldo() < 0) {
                    setEstadoJuego(EstadoJuego.ALGUNJUGADORENBANCARROTA);
                }
            }
        } else if (cartaActual.getTipo() == TipoSorpresa.CONVERTIRME) {
            Especulador especulador = jugadorActual.convertirme(cartaActual.getValor());//Preguntar

            int pos = jugadores.indexOf(jugadorActual);
            jugadores.remove(pos);
            jugadores.add(especulador);
            jugadorActual = especulador;
            System.out.println("-----------------AHORA ERES UN ESPECULADOR-------------");

        }

    }

    public boolean comprarTituloPropiedad() {
        boolean comprado = jugadorActual.comprarTituloPropiedad();
        if (comprado) {
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        }
        return comprado;

    }

    public boolean edificarCasa(int numeroCasilla) {
        boolean edificada = false;
        //1
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        //2
        TituloPropiedad titulo = casilla.getTitulo();
        //3
        edificada = jugadorActual.edificarCasa(titulo);
        if (edificada == true) {//hayEspacio && tengoSaldo
            //4
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        }
        return edificada;
    }

    public boolean edificarHotel(int numeroCasilla) {
        boolean edificada = false;
        //1
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        //2
        TituloPropiedad titulo = casilla.getTitulo();
        //3
        edificada = jugadorActual.edificarHotel(titulo);
        if (edificada == true) {//hayEspacio && tengoSaldo
            //4
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        }
        return edificada;
    }

    private void encarcelarJugador() {
        Casilla casillaCarcel = tablero.getCarcel();
        if (!jugadorActual.deboIrACarcel()) {

            jugadorActual.irACarcel(casillaCarcel);
            setEstadoJuego(EstadoJuego.JA_ENCARCELADO);
        } else {
            Sorpresa carta = jugadorActual.devolverCartaLibertad();
            mazo.add(carta);
            setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
        }
    }

    public boolean intentarSalirCarcel(MetodoSalirCarcel metodo) {
        if (metodo == MetodoSalirCarcel.TIRANDODADO) {
            int resultado = tirarDado();
            if (resultado >= 5) {
                jugadorActual.setEncarcelado(false);
            }
        } else if (metodo == MetodoSalirCarcel.PAGANDOLIBERTAD) {
            jugadorActual.pagarLibertad(PRECIO_LIBERTAD);
        }
        boolean libre = jugadorActual.getEncarcelado();
        if (libre) {
            setEstadoJuego(EstadoJuego.JA_ENCARCELADO);
        } else {
            setEstadoJuego(EstadoJuego.JA_PREPARADO);
        }
        return libre;
    }

    void mover(int numCasillaDestino) {
        Casilla casillaInicial = jugadorActual.getCasillaActual();
        Casilla casillaFinal = tablero.obtenerCasillaNumero(numCasillaDestino);
        jugadorActual.setCasillaActual(casillaFinal);
        if (numCasillaDestino < casillaInicial.getNumeroCasilla()) {
            jugadorActual.modificarSaldo(SALDO_SALIDA);
        }
        if (casillaFinal.soyEdificable()) {
            actuarSiEnCasillaEdificable();
        } else {
            actuarSiEnCasillaNoEdificable();
        }
    }

    public void hipotecarPropiedad(int numeroCasilla) {
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        TituloPropiedad titulo = casilla.getTitulo();
        jugadorActual.hipotecarPropiedad(titulo);
        setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);
    }

    public void venderPropiedad(int numeroCasilla) {

        //1
        Casilla casilla = tablero.obtenerCasillaNumero(numeroCasilla);
        //2
        jugadorActual.venderPropiedad(casilla);
        setEstadoJuego(EstadoJuego.JA_PUEDEGESTIONAR);

    }

    public boolean cancelarHipoteca(int numCasilla) {

        boolean cancelarHipoteca = false;
        Casilla casilla = tablero.obtenerCasillaNumero(numCasilla);
        TituloPropiedad titulo = casilla.getTitulo();
        boolean estaHipotecada = casilla.getTitulo().getHipotecada();
        boolean esDeMiPropiedad = jugadorActual.esDeMiPropiedad(titulo);
        boolean esEdificable = casilla.soyEdificable();
        if (esEdificable && esDeMiPropiedad && estaHipotecada) {
            cancelarHipoteca = jugadorActual.cancelarHipoteca(titulo);
        }

        estado = EstadoJuego.JA_PUEDEGESTIONAR;
        return cancelarHipoteca;

    }

}

File: QytetetRuby/lib/qytetet.rb
#encoding: utf-8

require_relative "tipo_sorpresa"
require_relative "sorpresa"
require_relative "jugador"
require_relative "dado"
require_relative "estado_juego"
require "singleton"
require_relative "especulador"
require_relative "calle"
require_relative "casilla"
require_relative "tablero"
require_relative "metodo_salir_carcel"

module ModeloQytetet
  class Qytetet
    include Singleton
    
    MAX_JUGADORES = 4
    NUM_SORPRESAS = 10 
    NUM_CASILLAS = 20
    PRECIO_LIBERTAD = 200
    SALDO_SALIDA = 1000
    
    attr_accessor :cartaActual,:estadoJuego
    attr_reader :mazo, :tablero, :dado, :jugadorActual, :jugadores
    
    
    def initialize
      @mazo = Array.new
      @cartaActual = nil
      @jugadores = Array.new
      @jugadorActual = nil
      @dado = Dado.instance
      @estadoJuego = nil
    end
    
    def actuarSiEnCasillaEdificable
      deboPagar = @jugadorActual.deboPagarAlquiler
      
      if(deboPagar)
        @jugadorActual.pagarAlquiler
        if(@jugadorActual.saldo <=0)
          @estadoJuego = (EstadoJuego::ALGUNJUGADORENBANCAROTA)
        end
      end
      
      casilla=obtenerCasillaJugadorActual
      tengoPropietario = casilla.tengoPropietario
      
      if(@estadoJuego != EstadoJuego::ALGUNJUGADORENBANCARROTA)
        if(tengoPropietario)
          @estadoJuego =(EstadoJuego::JA_PUEDEGESTIONAR)
        else
          @estadoJuego =(EstadoJuego::JA_PUEDECOMPRAROGESTIONAR)
        end
      end
    end
    
    def actuarSiEnCasillaNoEdificable
      @estadoJuego =(EstadoJuego::JA_PUEDEGESTIONAR)
      
      @casillaActual = @jugadorActual.casillaActual
      
      if(@casillaActual.tipo == TipoCasilla::IMPUESTO)
        @jugadorActual.pagarImpuesto
      else
        if (@casillaActual.tipo == TipoCasilla::JUEZ)
          encarcelarJugador
        else
          if(@casillaActual.tipo == TipoCasilla::SORPRESA)
            @cartaActual = @mazo.at(0)
            @mazo.delete(0)
            @estadoJuego =(EstadoJuego::JA_CONSORPRESA)
          end
        end
      end
    end
    def self.getMaxJugadores
      MAX_JUGADORES
    end
    def aplicarSorpresa
      @estadoJuego = EstadoJuego::JA_PUEDEGESTIONAR
      
      if(@cartaActual.tipo == TipoSorpresa::SALIRCARCEL)
        @jugadorActual.cartaLibertad = @cartaActual
      else
        @mazo<<@cartaActual
        
        if(@cartaActual.tipo == TipoSorpresa::PAGARCOBRAR)
          @jugadorActual.modificarSaldo(@cartaActual.valor)
          
          if(@jugadorActual.saldo < 0)
            @estadoJuego =(EstadoJuego::ALGUNJUGADORENBANCARROTA)
          end
        elsif (@cartaActual.tipo == TipoSorpresa::IRACASILLA)
          valor = @cartaActual.valor
          casillaCarcel = @tablero.esCasillaCarcel(valor)
          
          if(casillaCarcel)
            encarcelarJugador()
          else
            mover(valor)
          end
        elsif (@cartaActual.tipo == TipoSorpresa::PORCASAHOTEL)
          cantidad = @cartaActual.valor
          numeroTotal = @jugadorActual.cuantasCasasHotelesTengo
          @jugadorActual.modificarSaldo(cantidad*numeroTotal)
          
          if(@jugadorActual.saldo < 0)
            @estadoJuego = EstadoJuego::ALGUNJUGADORENBANCARROTA
          end
        elsif (@cartaActual.tipo == TipoSorpresa::PORJUGADOR)
          @jugadores.each do |jugador|
            if(jugador != @jugadorActual)
              jugador.modificarSaldo(@cartaActual.valor)
              
              if(jugador.saldo < 0)
                @estadoJuego = EstadoJuego::ALGUNJUGADORENBANCARROTA
              end
              
              jugador.modificarSaldo(-@cartaActual.valor)
              
              if(@jugadorActual.saldo < 0)
                @estadoJuego = EstadoJuego::ALGUNJUGADORENBANCARROTA
              end
            end
          end
        elsif @cartaActual.tipo == TipoSorpresa::CONVERTIRME
          
          especulador = @jugadorActual.convertirme(@cartaActual.valor) 
          indice = @jugadores.index(@jugadorActual)
          @jugadores[indice] = especulador
          @jugadorActual = especulador
          
          
        
        end
      end
    end
    
    def cancelarHipoteca(numeroCasilla)
      cancelarHipoteca = false
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      estaHipotecada = titulo.hipotecada
      esDeMiPropiedad = @jugadorActual.esDeMiPropiedad(titulo);
      esEdificable = casilla.soyEdificable
      if (esEdificable && esDeMiPropiedad && estaHipotecada) 
        cancelarHipoteca = @jugadorActual.cancelarHipoteca(titulo);
      end

      
      @estadoJuego = EstadoJuego::JA_PUEDEGESTIONAR
      
      return cancelarHipoteca
    end
    
   
    
    def comprarTituloPropiedad
      comprado = @jugadorActual.comprarTituloPropiedad
      
      if(comprado)
        @estadoJuego =(EstadoJuego::JA_PUEDEGESTIONAR)
      end
      
      return comprado
    end
    
    def edificarCasa(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      edificada = @jugadorActual.edificarCasa(titulo)
      
      if(edificada)
        @estadoJuego =(EstadoJuego::JA_PUEDEGESTIONAR)
      end
      
      return edificada
    end
    
    def edificarHotel(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      edificado = @jugadorActual.edificarHotel(titulo)
      
      if(edificado)
        @estadoJuego =(EstadoJuego::JA_PUEDEGESTIONAR)
      end
      
      return edificado
    end
    
    def encarcelarJugador
      if !@jugadorActual.tengoCartaLibertad
        casillaCarcel = @tablero.carcel
        @jugadorActual.irACarcel(casillaCarcel)
        @estadoJuego = EstadoJuego::JA_ENCARCELADO
        puts "Encarcelado"
      else
        carta = @jugadorActual.devolverCartaLibertad
        puts "Te libras de la carcel"
        @mazo << carta
        @estadoJuego = EstadoJuego::JA_PUEDEGESTIONAR
      end
    end
    
    def getValorDado
      return @dado.valor
    end
    
    def hipotecarPropiedad(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      titulo = casilla.titulo
      @jugadorActual.hipotecarPropiedad(titulo)
      @estadoJuego =(EstadoJuego::JA_PUEDEGESTIONAR)
    end
    
    def inicializarJuego(nombres)
      inicializarJugadores(nombres)
      inicializarTablero
      inicializarCartasSorpresa
      salidaJugadores
    end
    
    def inicializarJugadores(nombres)
      nombres.each do |i|
        @jugadores << Jugador.new(i)
      end
    end
    
    def inicializarTablero
      @tablero = Tablero.new
    end
    
    def intentarSalirCarcel(metodo)
      if(metodo == MetodoSalirCarcel::TIRANDODADO)
        resultado = tirarDado
        if(resultado >= 5)
          @jugadorActual.encarcelado = false
        end
      elsif (metodo == MetodoSalirCarcel::PAGANDOLIBERTAD)
        @jugadorActual.pagarLibertad(PRECIO_LIBERTAD)   
      end
      
      encarcelado = @jugadorActual.encarcelado
      if(encarcelado)
        @estadoJuego =(EstadoJuego::JA_ENCARCELADO)
      else
        @estadoJuego =(EstadoJuego::JA_PREPARADO) 
      end
      return (!encarcelado)
    end
    
    def jugar
      tirarDado
      casillaFinal = @tablero.obtenerCasillaFinal(@jugadorActual.casillaActual, @dado.valor)
      mover(casillaFinal.numeroCasilla)
    end
    
    def mover(numCasillaDestino)
      casillaInicial = @jugadorActual.casillaActual
      casillaFinal   = @tablero.obtenerCasillaNumero(numCasillaDestino)
      
      @jugadorActual.casillaActual = casillaFinal

      if(numCasillaDestino < casillaInicial.numeroCasilla)
        @jugadorActual.modificarSaldo(SALDO_SALIDA)
      end

      if(casillaFinal.soyEdificable)
        actuarSiEnCasillaEdificable
      else
        actuarSiEnCasillaNoEdificable
      end
    end
    
    def obtenerCasillaJugadorActual
      return @jugadorActual.casillaActual
    end
    
    
    
    def obtenerPropiedadesJugador
      numero_casillas_propiedades_jugador = Array.new
      propiedades = @jugadorActual.propiedades
      casillas = @tablero.casillas
      
      casillas.each do |c|
        propiedades.each do |p|
          if(c.titulo == p)
            numero_casillas_propiedades_jugador << c.numeroCasilla
          end
        end
      end
      
      return numero_casillas_propiedades_jugador
    end
    
    def obtenerPropiedadesJugadorSegunEstadoHipoteca(estadoHipoteca)
      n_casillas_jugador = Array.new
      propiedades = @jugadorActual.obtenerPropiedades(estadoHipoteca)
      
      @tablero.casillas.each do |c|
        propiedades.each do |p|
          if(c.titulo == p)
            n_casillas_jugador << c.numeroCasilla
          end
        end
      end
      
      return n_casillas_jugador
    end
    
    def obtenerRanking
      @jugadores=@jugadores.sort
    end
    
    def obtenerSaldoJugadorActual()
      return @jugadorActual.saldo
    end
    
    def salidaJugadores
      @jugadores.each do |j|
        j.casillaActual = @tablero.casillas.at(0)
      end
      primero = rand(0...@jugadores.length)
      @jugadorActual = @jugadores.at(primero)
      @estadoJuego = EstadoJuego::JA_PREPARADO
    end
    

    
   
    
    def siguienteJugador
      index = @jugadores.index(@jugadorActual)
      index = (index + 1)%@jugadores.size
      @jugadorActual = @jugadores.at(index)
      if (@jugadorActual.encarcelado)
        @estadoJuego = EstadoJuego::JA_ENCARCELADOCONOPCIONDELIBERTAD
      else
        @estadoJuego = EstadoJuego::JA_PREPARADO
      end
    end
    def tirarDado
      return @dado.tirar 
    end
    
    def venderPropiedad(numeroCasilla)
      casilla = @tablero.obtenerCasillaNumero(numeroCasilla)
      @jugadorActual.venderPropiedad(casilla)
      @estadoJuego = EstadoJuego::JA_PUEDEGESTIONAR
    end
   
    def inicializarCartasSorpresa
      @mazo.clear
      @mazo << Sorpresa.new("Un fan anónimo ha pagado tu fianza. Sales de la cárcel", 0, TipoSorpresa::SALIRCARCEL)
      @mazo << Sorpresa.new("Los killos te mandan de paseico", 1, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("Lo siento, pero te quitamos un pizquito de tu saldo", 2, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("Nos debes dinerico señor, ¡a pagar!", 3, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("Killo que te toca pagar, $$$ , MALO MARIA", 4, TipoSorpresa::PORJUGADOR)
      @mazo << Sorpresa.new("Killo que ahora si te llevas el pellizquito", 5, TipoSorpresa::PORJUGADOR)
      @mazo << Sorpresa.new("Los killos te mandan de paseico un ratejo por si no te ha tocado", 6, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("Fiuuuuuuu, que alegría que recibes dinerico", 7, TipoSorpresa::PAGARCOBRAR)
      @mazo << Sorpresa.new("Alegria señora que nos llevamos dinerico", 8, TipoSorpresa::PORCASAHOTEL)
      @mazo << Sorpresa.new("Te hemos pillado con chanclas y calcetines, lo sentimos,¡debes ir a la carcel!", 9, TipoSorpresa::IRACASILLA)
      @mazo << Sorpresa.new("Convertirseee!", 3000, TipoSorpresa::CONVERTIRME)
      @mazo << Sorpresa.new("Convertirseee!", 5000, TipoSorpresa::CONVERTIRME)
      @mazo.shuffle
    
    end 
    
    def to_s
      puts "--------------------- Juego Qytetet --------------------- \ncartaActual: #{@cartaActual} \nDado: #{@dado} \nJugador Actual: #{@jugadorActual}" 
      
      puts "\nJugadores:"
      
      @jugadores.each do |j|
        puts j
      end
      
      puts "\nMazo:" 
      @mazo.each do |m|
        puts m
      end
      
      
      "\nTablero #{@tablero}"  
    end
    def self.getMaxJugadores
      MAX_JUGADORES
    end
    private :encarcelarJugador, :inicializarCartasSorpresa, :inicializarJugadores, :inicializarTablero
    
  end
end
